File: src/DevProjects.Core/Services/ClaudeVersionInfo.cs
using System.Text.RegularExpressions;

namespace DevProjects.Core.Services;

/// <summary>Parses and compares claude CLI semver strings. Unknown input never nags.</summary>
public static partial class ClaudeVersionInfo
{
    [GeneratedRegex(@"(\d+)\.(\d+)\.(\d+)")]
    private static partial Regex Semver();

    public static (int Major, int Minor, int Patch)? Parse(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return null;
        var m = Semver().Match(raw);
        if (!m.Success) return null;
        return (int.Parse(m.Groups[1].Value), int.Parse(m.Groups[2].Value), int.Parse(m.Groups[3].Value));
    }

    public static bool IsOutdated(string? installed, string? latest)
    {
        var a = Parse(installed);
        var b = Parse(latest);
        if (a is null || b is null) return false;
        return Compare(b.Value, a.Value) > 0;
    }

    private static int Compare((int Major, int Minor, int Patch) x, (int Major, int Minor, int Patch) y)
    {
        if (x.Major != y.Major) return x.Major.CompareTo(y.Major);
        if (x.Minor != y.Minor) return x.Minor.CompareTo(y.Minor);
        return x.Patch.CompareTo(y.Patch);
    }
}

File: electron/src/core/claude/claudeVersionInfo.ts
/** Parses and compares claude CLI semver strings. Unknown input never nags. */

export interface SemVer {
  major: number
  minor: number
  patch: number
}

const SEMVER_RE = /(\d+)\.(\d+)\.(\d+)/

/** Extracts the first semver triple from raw, or null if not found / blank. */
export function parse(raw: string | null | undefined): SemVer | null {
  if (!raw || !raw.trim()) return null
  const m = SEMVER_RE.exec(raw)
  if (!m) return null
  return { major: parseInt(m[1], 10), minor: parseInt(m[2], 10), patch: parseInt(m[3], 10) }
}

/** Returns true when latest is strictly newer than installed; false for any unknown. */
export function isOutdated(installed: string | null | undefined, latest: string | null | undefined): boolean {
  const a = parse(installed)
  const b = parse(latest)
  if (!a || !b) return false
  return compareSemver(b, a) > 0
}

function compareSemver(x: SemVer, y: SemVer): number {
  if (x.major !== y.major) return x.major - y.major
  if (x.minor !== y.minor) return x.minor - y.minor
  return x.patch - y.patch
}
